README

                      Template Toolkit LaTeX Plugin

                            Version 3.00_04

                           13 October 2007

        Copyright (C) 2006-2007 Andrew Ford.   All Rights Reserved
        Copyright (C) 1996-2007 Andy Wardley.  All Rights Reserved

          This is free software; you can redistribute it and/or
             modify it under the same terms as Perl itself.


DESCRIPTION
-----------

The Template Toolkit is a collection of modules which implement a
fast, flexible, powerful and extensible template processing system.
The LaTeX plugin provides the facility to filter content through
LaTeX, provided of course that the content is valid LaTeX source.  

The LaTeX plugin takes care of running and re-running latex so that
forward references, tables of contents, and lists of figures and
tables are resolved.  It will also run bibtex and makeindex if it
detects that a bibliography or in index have been specified, and will
re-run latex again one or more times until the formatting of the
document has stabilized.  (This functionality is new in version 3.00
of the module.)



QUICK START
-----------

The latest version of the Template Toolkit LaTeX Plugin can be
retrieved from:

    http://www.cpan.org/modules/by-module/Template-Latex/

Fetch and install Template Toolkit (2.16 or newer) if you don't
already have it installed.  Available from CPAN in:

    http://www.cpan.org/authors/Andy_Wardley/

To install the Template Toolkit Latex Plugin:

    tar zxf Template-Latex-3.00_04.tar.gz
    cd Template-Latex-3.00_04
    perl Makefile.PL
    make
    make test
    make install    

The Makefile.PL will prompt for additional configuration options,
including the installation of optional template libraries, HTML
documentation and examples.  You can safely answer 'n' to all of these
questions for a quick and basic installation.  

If you're running ActivePerl on a Win32 platform then you can use the
Perl Package Manager (PPM) to install the Template Toolkit.  Chris
Winters maintains a repository of pre-compiled PPM packages which contains
the Template Toolkit, AppConfig and others.  For further information, see:

    http://openinteract.sourceforge.net/


For further details on installation, see the separate INSTALL file.


WHAT'S NEW?
-----------

Version 3.00 is a complete rewrite of Template-Latex.

Version 2.16 was the first release of Template-Latex.  It was created
by moving the code out of the core Template Toolkit distribution
(version 2.15).

See the Changes file for further details of the changes in these
releases.


GENERAL FEATURES
----------------

Some of the key features of the Template Toolkit Latex Plugin are listed below.
See the documentation for further detail.


DOCUMENTATION
-------------

The Template Toolkit is provided with enough documentation to keep all
but the most voracious reader happy for quite some time.

The 'Changes' file in the distribution directory documents all visible
changes between versions of the Template Toolkit.  See the section 
'VERSION COMPATABILITY' below for further details.

The 'TODO' file, also in the distribution directory, lists known bugs,
planned enhancements and possible new features for future versions.

The rest of the documentation is distributed in Pod and HTML formats.
The Pod pages are installed when you 'make install' and can be viewed
using 'perldoc', e.g.


SUPPORT
-------

The Template Toolkit mailing list provides a forum for discussing
issues relating to the use and abuse of the Template Toolkit.  There
are a number of knowledgeable and helpful individuals who frequent the
list (including the author) who can often offer help or suggestions.
Please respect their time and patience by checking the documentation
and/or mailing list archives before asking questions that may already
have been answered.   

To subscribe to the mailing list, send an email to:

    templates-request@template-toolkit.org

with the message 'subscribe' in the body.  You can also use the web 
interface to subscribe or browse the archives:

    http://www.template-toolkit.org/mailman/listinfo/templates

A low-volume, moderated mailing list exists for announcements about new
releases of the Template Toolkit and related products.  To subscribe, send
an email to:

    templates-announce-request@template-toolkit.org

with the message 'subscribe' in the body.  A web interface also exists
for subscription and browsing the archives:

    http://www.template-toolkit.org/mailman/listinfo/templates-announce

For information about commercial support and consultancy for the Template 
Toolkit, please contact the author.


AUTHOR
------

The Template Toolkit Latex plugin was originally written as part of
the Template Toolkit distribution by Andy Wardley <abw@wardley.org>
with the assistance and contributions from many other people. In 2006
the Latex plugin was separated out into its own distribution and
Andrew Ford <a.ford@ford-mason.co.uk> the took over as maintainer.

See Template::Manual::Credits for details.


COPYRIGHT
---------

Copyright (C) 2006-2007 Andrew Ford.   All Rights Reserved.
Copyright (C) 1996-2007 Andy Wardley.  All Rights Reserved.

This is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.



Changes
#================================================= -*- indented-text -*-
#
# Changes 
#
# DESCRIPTION
#   Revision history for the Template Toolkit Latex plugin.
#
# AUTHOR
#   Andrew Ford   <A.Ford@ford-mason.co.uk>
#
#
#------------------------------------------------------------------------
# $Id: Changes 86 2009-03-11 17:45:23Z ford $
#========================================================================

#------------------------------------------------------------------------
# Version 3.06
# -----------------------------------------------------------------------

* Makefile now generates without user input
* Test cases should now run on all smoketesters.

#------------------------------------------------------------------------
# Version 3.03 to 03.05
#------------------------------------------------------------------------

* Minor packaging fixes which correct issues relating to CPAN indexing as
  a co-committer.  No code changes.

#------------------------------------------------------------------------
# Version 3.02.01 - 2014-02-25
#------------------------------------------------------------------------

* Fixed test case failures based on differing escape returns from
  LaTeX::Encode

#------------------------------------------------------------------------
# Version 3.02 - 2009-03-12
#------------------------------------------------------------------------

* fixed skip_all call (use Template::Test rather than Test::More - 
  despite Template::Test being incompatible with TAP)

#------------------------------------------------------------------------
# Version 3.01 - 2009-03-11
#------------------------------------------------------------------------

* tidied up tests to avoid spurious failures (skip_all if the 'dvitype'
  program is not available)

* expanded documentation

#------------------------------------------------------------------------
# Version 3.00 - 2009-03-09
#------------------------------------------------------------------------

* first non-developer release - no changes from 3.00_05


#------------------------------------------------------------------------
# Version 3.00_05 - 2007-10-13
#------------------------------------------------------------------------

* Removed "use Template" from Makefile.PL as that was unnecessary and
  was causing build failures where Template was not already installed.


#------------------------------------------------------------------------
# Version 3.00_04 - 2007-10-13
#------------------------------------------------------------------------

* Added a test script for the "table" method

* Updated dependency on LaTeX::Driver to 0.07 as 0.06 failed to build
  on cpantesters machines

* mark 11-plugin-errors.t as all skipped, as the error messages that
  are checked for are not what the driver module produces - I need to
  review this test script and change some of the individual checks to
  do meaningful things

#------------------------------------------------------------------------
# Version 3.00_03 - 2007-10-03
#------------------------------------------------------------------------

* Added a "table" method to the plugin to interface to the
  LaTeX::Table module.  This will generate the LaTeX commands to
  format a table from raw data.

#------------------------------------------------------------------------
# Version 3.00_02 - 2007-10-03
#------------------------------------------------------------------------

* Restructured again so that Template::Latex calls the
  Template::Plugin::Latex module, which in turn builds on
  LaTeX::Driver to format the document.  (Removed the
  Template::Latex::Paths module as the paths are now held in the
  LaTeX::Driver module.)

* Expanded the test suite and the documentaton.

#------------------------------------------------------------------------
# Version 3.00_01 - 2007-09-13
#------------------------------------------------------------------------

* I have restructured the plugin fairly extensively, and added
  radically new features, so this release has a new major version
  number, i.e. 3.x.

* Restructured Template::Latex to run latex multiple times if
  necessary and also to run bibtex if citations are detected and
  makeindex if indexing is detected.

* The TEXINPUTS environment variable is set up to include the original
  directory so that include files can be found there.

* Change the format of the commands passed to system() so that the cd
  into the temporary directory is done in the command rather than in
  the module.  That way the module does not need to change back to the
  original directory, simplifying error handling.  
  (The Win32 idiom is: 'cmd /c "cd DIR && COMMAND"')

* Split the driver code in Template::Latex out into its own module
  (Template::Latex::Driver).

* The LaTeX macro \TTLATEX is defined when latex or pdflatex is run

* The static filter detex is defined by the plugin

#------------------------------------------------------------------------
# Version 2.17 - 2006-06-07
#------------------------------------------------------------------------

* Changes dependency on Template to version 2.15 as version 2.16 of
  Template, which was specified as the dependency has not been relased.

* Added files: Changes (this file) and TODO


#------------------------------------------------------------------------
# Version 2.16 - 2006-05-31
#------------------------------------------------------------------------

First release of the plugin as a separate distribution to the core
Template Toolkit distribution.  New maintainer: Andrew Ford.


t/lib/Template/Test/Latex.pm
package Template::Test::Latex;

use strict;
use vars qw(@ISA @EXPORT);
use Config;
use Test::More;

require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(find_program grep_dvi dvitype require_dvitype);

our $WIN32  = ($^O eq 'MSWin32');

$Template::Plugin::Latex::DEBUG = grep(/-d/, @ARGV);
$Template::Plugin::Latex::DEBUG = $Template::Plugin::Latex::DEBUG;

my $dvitype = find_program($ENV{PATH}, "dvitype");

sub dvitype {
    return $dvitype;
}

sub require_dvitype {
    if (!$dvitype || ! -x $dvitype) {
        plan skip_all => "'dvitype' is not available";
        exit(0);
    }
}


#------------------------------------------------------------------------
# find_program($path, $prog)
#
# Find a program, $prog, by traversing the given directory path, $path.
# Returns full path if the program is found.
#
# Written by Craig Barratt, Richard Tietjen add fixes for Win32.
#
# abw changed name from studly caps findProgram() to find_program() :-)
#------------------------------------------------------------------------

sub find_program {
    my($path, $prog) = @_;

    foreach my $dir ( split($Config{path_sep}, $path) ) {
        my $file = File::Spec->catfile($dir, $prog);
        if ( !$WIN32 ) {
            return $file if ( -x $file );
        } else {
            # Windows executables end in .xxx, exe precedes .bat and .cmd
            foreach my $dx ( qw/exe bat cmd/ ) {
                return "$file.$dx" if ( -x "$file.$dx" );
            }
        }
    }

}

sub grep_dvi {
    my $dir = shift;
    my $file = shift;
    my $regexp = shift;
    my $path = File::Spec->catfile($dir, $file);
    return "FAIL - $file does not exist" unless -f $path;
    return "FAIL - can't find dvitype" unless $dvitype and -x $dvitype;
    my $dvioutput =  `$dvitype $path`; 
    foreach (split(/\n/, $dvioutput)) {
	next unless /^\[(.*)\]$/;
	return "PASS - found '$regexp'" if /$regexp/;
    }
    return "FAIL - '$regexp' not found";
}

1;


lib/Template/Latex/Paths.pm
#============================================================= -*-perl-*-
#
# Template::Latex::Paths
#
# DESCRIPTION
#   Provides an interface to Latex from the Template Toolkit.
#
# AUTHOR
#   Andrew Ford    <a.ford@ford-mason.co.uk>  (current maintainer)
#   Andy Wardley   <abw@wardley.org>
#
# COPYRIGHT
#   Copyright (C) 1996-2006 Andy Wardley.  All Rights Reserved.
#
#   This module is free software; you can redistribute it and/or
#   modify it under the same terms as Perl itself.
#
# HISTORY
#   * Latex plugin originally written by Craig Barratt, Apr 28 2001.
#   * Win32 additions by Richard Tietjen.
#   * Extracted into a separate Template::Latex module by Andy Wardley,
#     May 2006
#   * Removed the functionality to specify program pathnames on the FILTER call
#     Andrew Ford, 05 June 2006
#
#========================================================================
 
package Template::Latex::Paths;

use strict;
use warnings;

# LaTeX executable paths set at installation time by the Makefile.PL
our $LATEX     = '/usr/bin/latex';
our $PDFLATEX  = '/usr/bin/pdflatex';
our $DVIPS     = '/usr/bin/dvips';
our $PS2PDF    = '/usr/bin/ps2pdf';
our $BIBTEX    = '/usr/bin/bibtex';
our $MAKEINDEX = '/usr/bin/makeindex';

1;


lib/Template/Plugin/Latex.pm
#============================================================= -*-perl-*-
#
# Template::Plugin::Latex
#
# $Id:$
#
# DESCRIPTION
#   Template Toolkit plugin for Latex
#
# AUTHOR
#   Andrew Ford    <a.ford@ford-mason.co.uk>  (current maintainer)
#   Andy Wardley   <abw@wardley.org>          (original author)
#
# COPYRIGHT
#   Copyright (C) 2006-2007 Andrew Ford.   All Rights Reserved.
#   Copyright (C) 1996-2006 Andy Wardley.  All Rights Reserved.
#
#   This module is free software; you can redistribute it and/or
#   modify it under the same terms as Perl itself.
#
# HISTORY
#   * Originally written by Craig Barratt, Apr 28 2001.
#   * Win32 additions by Richard Tietjen.
#   * Extracted into a separate Template::Plugin::Latex module by 
#     Andy Wardley, 27 May 2006
#   * Removed the program pathname options on the FILTER call
#     Andrew Ford, 05 June 2006
#   * Totally rewritten by Andrew Ford September 2007
#   * Version 3.00 released March 2009
#
#========================================================================

package Template::Plugin::Latex;

use strict;
use warnings;
use base 'Template::Plugin';

use File::Spec;
use LaTeX::Driver 0.07;
use LaTeX::Encode;
use LaTeX::Table;


our $VERSION = 3.06;           # Update "=head1 VERSION" below!!!!
our $DEBUG; $DEBUG = 0 unless defined $DEBUG;
our $ERROR   = '';
our $FILTER  = 'latex';
our $THROW   = 'latex';        # exception type

#------------------------------------------------------------------------
# constructor
#
#------------------------------------------------------------------------

sub new {
    my ($class, $context, $options) = @_;

    # make sure that $options is a hash ref
    $options ||= {};

    # create a closure to generate filters with additional options
    my $filter_factory = sub {
        my $factory_context = shift;
        my $filter_opts = ref $_[-1] eq 'HASH' ? pop : { };
        my $filter_args = [ @_ ];
        @$filter_opts{ keys %$options } = values %$options;
        return sub {
            # Template::Plugin::Latex::_filter->run($context, $filter_opts, $filter_args, @_);
            _tt_latex_filter($class, $factory_context, $filter_opts, $filter_args, @_);
        };
    };

    # create a closure to generate filters with additional options
    my $encode_filter_factory = sub {
        my $factory_context = shift;
        my $filter_opts = ref $_[-1] eq 'HASH' ? pop : { };
        my $filter_args = [ @_ ];
        @$filter_opts{ keys %$options } = values %$options;
        return sub {
            # Template::Plugin::Latex::_filter->run($context, $filter_opts, $filter_args, @_);
            _tt_latex_encode_filter($class, $factory_context, $filter_opts, $filter_args, @_);
        };
    };

    # and a closure to represent the plugin
    my $plugin = sub {
        my $plugopt = ref $_[-1] eq 'HASH' ? pop : { };
        @$plugopt{ keys %$options } = values %$options;
        # Template::Plugin::Latex::_filter->run($context, $plugopt, @_ );
        _tt_latex_filter($class, $context, $plugopt, {}, @_ );
    };


    # now define the filter and return the plugin
    $context->define_filter('latex_encode', [ $encode_filter_factory => 1 ]);
    $context->define_filter($options->{filter} || $FILTER, [ $filter_factory => 1 ]);

    return bless $plugin, $class;
}


#------------------------------------------------------------------------
# _tt_latex_encode_filter
#
#
#------------------------------------------------------------------------

sub _tt_latex_encode_filter {
    my ($class, $context, $options, $filter_args, @text) = @_;
    my $text = join('', @text);
    return latex_encode($text, %{$options});
}


#------------------------------------------------------------------------
# _tt_latex_filter
#
#
#------------------------------------------------------------------------

sub _tt_latex_filter {
    my ($class, $context, $options, $filter_args, @text) = @_;
    my $text = join('', @text);

    # Get the output and format options

#    my $output = $options->{output};
    my $output = delete $options->{ output } || shift(@$filter_args) || '';
    my $format = $options->{format};

    # If the output is just a format specifier then set the format to
    # that and undef the output

    if ($output =~ /^ (?: dvi | ps | pdf(?:\(\w+\))? ) $/x) {
        ($format, $output) = ($output, undef);
    }

    # If the output is a filename then convert to a full pathname in
    # the OUTPUT_PATH directory, outherwise set the output to a
    # reference to a temporary variable.

    if ($output) {
        my $path = $context->config->{ OUTPUT_PATH }
            or $class->_throw('OUTPUT_PATH is not set');
        $output = File::Spec->catfile($path, $output);
    }
    else {
        my $temp;
        $output = \$temp;
    }

    # Run the formatter

    eval {
        my $drv = LaTeX::Driver->new( source    => \$text,
                                      output    => $output,
                                      format    => $format,
                                      maxruns   => $options->{maxruns},
                                      extraruns => $options->{extraruns},
                                      texinputs => _setup_texinput_paths($context),
                                    );
        $drv->run;
    };
    if (my $e = LaTeX::Driver::Exception->caught()) {
        $class->_throw("$e");
    }

    # Return the text if it was output to a scalar variable, otherwise
    # return nothing.

    return ref $output ? $$output : '';
}


#------------------------------------------------------------------------
# $self->setup_texinput_paths
#
# setup the TEXINPUT path environment variables
#------------------------------------------------------------------------

sub _setup_texinput_paths {
    my ($context) = @_;
    my $template_name = $context->stash->get('template.name');
    my $include_path  = $context->config->{INCLUDE_PATH} || [];
    $include_path = [ $include_path ] unless ref $include_path;

    my @texinput_paths = ("");
    foreach my $path (@$include_path) {
        my $template_path = File::Spec->catfile($path, $template_name);
        if (-f $template_path) {
            my($volume, $dir) = File::Spec->splitpath($template_path);
            $dir = File::Spec->catfile($volume, $dir);
            unshift @texinput_paths, $dir;
            next if $dir eq $path;
        }
        push @texinput_paths, $path;
    }
    return  \@texinput_paths;
}


sub _throw {
    my $self = shift;
    die Template::Exception->new( $THROW => join('', @_) );
}


sub table {
    my $args  = ref($_[-1]) eq 'HASH' ? pop(@_) : { };
    my ($table, $text);
    eval {
        $table = LaTeX::Table->new($args);
        $text  = $table->generate_string;
    };
    if ($@) {
        die Template::Exception->new( $THROW => $@ );
    }
    return $text;
}


1;

__END__

=head1 NAME

Template::Plugin::Latex - Template Toolkit plugin for Latex

=head1 VERSION

This documentation refers to C<Template::Plugin::Latex> version 3.06

=head1 SYNOPSIS

Sample Template Toolkit code:

    [%- USE Latex;

        mystr = "a, b & c" | latex_encode;

        FILTER latex("pdf");  -%]

    \documentclass{article}
    \begin{document}
      This is a PDF document generated by
      LaTeX and the Template Toolkit, with some
      interpolated data: [% mystr %]
    \end{document}

    [%  END; -%]


=head1 DESCRIPTION

The C<Latex> Template Toolkit plugin provides a C<latex> filter that
allows the use of LaTeX to generate PDF, PostScript and DVI output files
from the Template Toolkit.  The plugin uses L<LaTeX::Driver> to run the
various LaTeX programs.

Processing of the LaTeX document takes place in a temporary directory
that is deleted once processing is complete.  The standard LaTeX
programs (C<latex> or C<pdflatex>, C<bibtex> and C<makeindex>) are run
and re-run as necessary until all references, indexes, bibliographies,
table of contents, and lists of figures and tables are stable or it is
apparent that they will not stabilize.  The format converters C<dvips>,
C<dvipdf>, C<ps2pdf> and C<pdf2ps> are run as necessary to convert the
output document to the requested format.  The C<TEXINPUTS> environment
variable is set up to include the template directory and the C<INCLUDES>
directories, so that LaTeX file inclusion commands should find the
intended files.

The output of the filter is binary data (although PDF and PostScript are
not stictly binary).  You should be careful not to prepend or append any
extraneous characters (even space characters) or text outside the FILTER
block as this text will be included in the file output.  Notice in the
example below how we use the post-chomp flags ('-') at the end of the
C<USE> and C<END> directives to remove the trailing newline characters:

    [% USE Latex(format='pdf') -%]
    [% FILTER latex %]
    ...LaTeX document...
    [% END -%]

If you're redirecting the output to a file via the third argument of
the Template module's C<process()> method then you should also pass
the C<binmode> parameter, set to a true value to indicate that it is a
binary file.

    use Template;

    my $tt = Template->new({
        INCLUDE_PATH => '/path/to/templates',
        OUTPUT_PATH  => '/path/to/pdf/output',
    });
    my $vars = {
        title => 'Hello World',
    }
    $tt->process('example.tt2', $vars, 'example.pdf', binmode => 1)
        || die $tt->error();

If you want to capture the output to a template variable, you can do
so like this:

    [% output = FILTER latex %]
    ...LaTeX document...
    [% END %]

You can pass additional arguments when you invoke the filter, for
example to specify the output format.

    [% FILTER latex(format='pdf') -%]
       ...LaTeX document...
    [% END %]

If you want to write the output to a file then you can specify an
C<output> parameter.

    [% FILTER latex(output='example.pdf') %]
    ...LaTeX document...
    [% END %]

If you don't explicity specify an output format then the filename
extension (e.g. 'pdf' in the above example) will be used to determine
the correct format.

You can specify a different filter name using the C<filter> parameter.

    [% USE Latex(filter='pdf') -%]
    [% FILTER pdf %]
    ...LaTeX document...
    [% END %]

You can also specify the default output format.  This value can be
C<latex>, C<pdf> or C<dvi>.

    [% USE Latex(format='pdf') %]


Note: the C<LaTeX::Driver> distribution includes three filter programs
(C<latex2dvi>, C<latex2pdf> and C<latex2ps>) that use the
C<LaTeX::Driver> package to process LaTeX source data into DVI, PDF or
PostScript file respectively.  These programs have a C<-tt2> option to
run their input through the Template Toolkit before processing as LaTeX
source.  The programs do not use the C<Latex> plugin unless the template
requests it, but they may provide an alternative way of processing
Template Toolkit templates to generate typeset output.


=head1 SUBROUTINES/METHODS

=head2 C<USE Latex(options)>

This statement loads the plugin (note that prior to version 2.15 the
filter was built in to Template Toolkit so this statement was
unnecessary; it is now required).


=head2 The C<latex> Filter

The C<latex> filter accepts a number of options, which may be
specified on the USE statement or on the filter invocation.

=over 4

=item C<format>

specifies the format of the output; one of C<dvi> (TeX device
independent format), C<ps> (PostScript) or C<pdf> (Adobe Portable
Document Format).  The follow special values are also accepted:
C<pdf(ps)> (generates PDF via PostScript, using C<dvips> and
C<ps2pdf>), C<pdf(dvi)> (generates PDF via dvi, using C<dvipdfm>)

=item C<output>

the name of the output file, or just the output format

=item C<indexstyle>

the name of the C<makeindex> style file to use (this is passed with
the C<-s> option to C<makeindex>)

=item C<indexoptions>

options to be passed to C<makeindex>.  Useful options are C<-l> for
letter ordering of index terms (rather than the default word
ordering), C<-r> to disable implicit page range formation, and C<-c>
to compress intermediate blanks in index keys. Refer to L<makeindex(1)>
for full details.

=item C<maxruns>

The maximum number of runs of the formatter program (defaults to 10).

=item C<extraruns>

The number of additional runs of the formatter program after it seems
that the formatting of the document has stabilized (default 0).  Note
that the setting of C<maxruns> takes precedence, so if C<maxruns> is
set to 10 and C<extraruns> is set to 3, and formatting stabilizes
after 8 runs then only 2 extra runs will be performed.

=back


=head2 The C<latex_encode> filter

The C<latex_encode> filter encodes LaTeX special characters in its
input into their LaTeX encoded representations.  It also encodes other characters that have

The special characters are: C<\> (command character), C<{> (open
group), C<}> (end group), C<&> (table column separator), C<#>
(parameter specifier), C<%> (comment character), C<_> (subscript),
C<^> (superscript), C<~> (non-breakable space), C<$> (mathematics mode).


=over 4

=item C<except>

Lists the characters that should be excluded from encoding.  By
default no special characters are excluded, but it may be useful to
specify C<except = "\\{}"> to allow the input string to contain LaTeX
commands such as C<"this is \textbf{bold} text">.

=item C<use_textcomp>

By default the C<latex_encode> filter will encode characters with the
encodings provided by the C<textcomp> LaTeX package (for example the
Pounds Sterling symbol is encoded as C<\textsterling{}>).  Setting
C<use_textcomp = 0> turns off these encodings.

=back

=head2 C<table()>

The C<table()> function provides an interface to the C<LaTeX::Table> module.

The following example shows how a simple table can be set up.

    [%- USE Latex;

        data = [ [ 'London', 'United Kingdom' ],
                 [ 'Berlin', 'Germany'        ],
                 [ 'Paris',  'France'         ],
                 [ 'Washington', 'USA'        ] ] );

        text = Latex.table( caption  = 'Capitol Cities',
                            label    = 'table:capitols',
                            headings = [ [ 'City', 'Country' ] ],
                            data     = data );
     -%]

The variable C<text> will hold the LaTeX commands to typeset the table
and can be further interpolated into a LaTeX document template.

=head1 DIAGNOSTICS

Most failures result from invalid LaTeX input and are propogated up from
L<LaTeX::Driver>, L<LaTeX::Encode> or L<LaTeX::Table>.

Failures detected in this module include:

=over 4

=item C<OUTPUT_PATH is not set>

an output filename was specified but the C<OUTPUT_PATH> configuration
option has not been set.

=back


=head1 DEPENDENCIES

=over 4

=item L<Template>

The Template Toolkit.

=item L<LaTeX::Driver>

Provides the logic for driving the LaTeX programs.

=item L<LaTeX::Encode>

Underpins the C<latex_encode> filter.

=item L<LaTeX::Table>

Underpins the C<table> function.

=back


=head1 INCOMPATIBILITIES

The C<latex> filter was distributed as part of the core
Template Toolkit distribution until version 2.15 (released in May 2006),
when it was moved into the separate Template-Latex distribution.  The
C<Latex> plugin must now be explicity to enable the C<latex> filter.


=head1 BUGS AND LIMITATIONS

The paths to the F<latex>, F<pdflatex> and F<dvips> should be
pre-defined as part of the installation process of L<LaTeX::Driver>
(i.e. when you run C<perl Makefile.PL> for that package).  Alternative
values can be specified from Perl code using the C<program_path> class
method from that package, but there are deliberately no options to
specify these paths from TT code.


=head1 AUTHOR

Andrew Ford E<lt>a.ford@ford-mason.co.ukE<gt> (current maintainer)

Andy Wardley E<lt>abw@wardley.orgE<gt> L<http://wardley.org/>

The original Latex plugin on which this is based was written by Craig
Barratt with additions for Win32 by Richard Tietjen.  The code has
subsequently been radically refactored by Andrew Ford.

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2006-2009 Andrew Ford.  All Rights Reserved.

Copyright (C) 1996-2006 Andy Wardley.  All Rights Reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

This software is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=head1 SEE ALSO

L<Template>, L<LaTeX::Driver>, L<LaTeX::Table>, L<LaTeX::Encode>


L<latex2dvi(1)>, L<latex2pdf(1)> and L<latex2ps(1)> (part of the
C<LaTeX::Driver> distribution)

=cut


# Local Variables:
# mode: perl
# perl-indent-level: 4
# indent-tabs-mode: nil
# End:
#
# vim: expandtab shiftwidth=4:


Makefile.PL
#!/usr/bin/perl -w                                         # -*- perl -*-

use strict;

use lib qw( ./lib );
use Config;
use File::Spec::Functions qw( catfile );
use Template;
use ExtUtils::MakeMaker;
use Cwd;

select STDERR;
$| = 1;
select STDOUT;

our $QUIET  = 0;
our $ACCEPT = 0;
our $WIN32  = ($^O eq 'MSWin32');
our $MODVERSION = get_module_version();

#------------------------------------------------------------------------

message(<<EOF);

Template-Plugin-Latex v$MODVERSION
------------------------------

The Template-Latex modules add LaTeX support for the Template Toolkit.

They allow you to create LaTeX documents using the full power of the
Template Toolkit.  The generated output can then be processed with
either 'pdflatex' or 'latex'.  If unresolved cross references,
bibliographic references or index definitions are found then 'bibtex'
or 'makeindex' will be run as appropriate and 'latex' or 'pdflatex'
re-run as necessary.  The output will be postprocessed with the
'dvips' and 'ps2pdf' programs if necessary to create PDF, DVI or
PostScript documents.

To use the Template-Latex module you will first need to install LaTeX on
your system and make sure the above programs are available.  

EOF

#------------------------------------------------------------------------

if (! find_program($ENV{PATH}, "latex") and ! find_program($ENV{PATH}, "pdflatex")) {
    my $continue = ttprompt('You don\'t seem to have LaTeX installed.  Continue anyway?', 'N');
    die("Aborting as LaTeX is not installed\n") unless uc($continue) =~ /^Y/;
}


#------------------------------------------------------------------------

my %opts = (
    'NAME'	       => 'Template-Plugin-Latex',
    'VERSION_FROM' => 'lib/Template/Plugin/Latex.pm',
    'PMLIBDIRS'    => [ 'lib' ], 
    'PREREQ_PM'    => { 
        'Template'      => 2.16,
	'LaTeX::Driver' => 0.07,
	'LaTeX::Encode' => 0.02,
	'LaTeX::Table'  => 0,
    },
    'dist'         => {
        'COMPRESS' => 'gzip',
        'SUFFIX'   => 'gz',
    },
    'clean'        => {
        'FILES'    => join(' ', qw( t/output/test1.pdf 
                                    t/output/test1.ps 
                                    t/output/test1.dvi
                                    t/output/test2 
                                    t/output/test2.pdf 
                                    t/output/test2.ps 
                                    t/output/test2.dvi )),
    },
);

if ($ExtUtils::MakeMaker::VERSION >= 5.43) {
    $opts{ AUTHOR   } = 'Andrew Ford <a.ford@ford-mason.co.uk>';
    $opts{ ABSTRACT } = 'Latex support for the Template Toolkit',
}

WriteMakefile(%opts);



#------------------------------------------------------------------------
# find_program($path, $prog)
#
# Find a program, $prog, by traversing the given directory path, $path.
# Returns full path if the program is found.
#
# Written by Craig Barratt, Richard Tietjen add fixes for Win32.
#
# abw changed name from studly caps findProgram() to find_program() :-)
#------------------------------------------------------------------------

sub find_program {
    my($path, $prog) = @_;

    foreach my $dir ( split($Config{path_sep}, $path) ) {
        my $file = File::Spec->catfile($dir, $prog);
        if ( !$WIN32 ) {
            return $file if ( -x $file );
        } else {
            # Windows executables end in .xxx, exe precedes .bat and .cmd
            foreach my $dx ( qw/exe bat cmd/ ) {
                return "$file.$dx" if ( -x "$file.$dx" );
            }
        }
    }
}


#------------------------------------------------------------------------
# message($text)
#
# Print message unless quiet mode.
#------------------------------------------------------------------------

sub message {
    return if $QUIET;
    print @_;
}


#------------------------------------------------------------------------
# ttprompt($message, $default)
#------------------------------------------------------------------------

sub ttprompt {
    my ($msg, $def)=@_;
    my $ISA_TTY = -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT)) ; # Pipe?
    my $dispdef = defined $def ? "[$def] " : " ";
    $def = defined $def ? $def : "";
    my $ans = '';
    local $|=1;
    print "$msg $dispdef" unless $QUIET;
    if ($ACCEPT || ! $ISA_TTY) {
        print "$def\n" unless $QUIET;
    }
    else {
        chomp($ans = <STDIN>);
    }
    return ($ans ne '') ? $ans : $def;
}


#------------------------------------------------------------------------
# get_module_version()
#
# Reads the module version from the Template::Plugin::Latex module file
#------------------------------------------------------------------------

sub get_module_version {
    if (open(MODULE, catfile('lib','Template','Plugin','Latex.pm'))) {
	while (<MODULE>) {
	    return $1 if /VERSION = "?(\d+\.[\d_]+)/;
	}
    }
    return "<unkown version>";
}
